core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -161,14 +237,26 @@
     private final Set<Class<? extends PipelineAggregator.Parser>> pipelineAggParsers = new HashSet<>();
     private final Highlighters highlighters = new Highlighters();
     private final Suggesters suggesters = new Suggesters();
-    private final Set<Class<? extends ScoreFunctionParser>> functionScoreParsers = new HashSet<>();
+    private final Map<String, ScoreFunctionParser<?>> functionScoreParsers = new HashMap<>();
+    private final List<Supplier<QueryParser<?>>> queryParsers = new ArrayList<>();
```

Comment:
for my education, what is the advantage of using a method reference for query parsers over passing in an actual instance of the query parser? If there is an advantage why don't we do the same for ScoreFunctionParsers too?